packages/nextra/src/server/rehype-plugins/twoslash/rehype-twoslash-popup.ts

```diff
@@ -0,0 +1,140 @@
+import { rendererRich, type RendererRichOptions } from '@shikijs/twoslash'
+import type { Element, ElementContent, Root } from 'hast'
+import type { Code } from 'mdast'
+import { fromMarkdown } from 'mdast-util-from-markdown'
+import { gfmFromMarkdown } from 'mdast-util-gfm'
+import { defaultHandlers, toHast } from 'mdast-util-to-hast'
+import type { ShikiTransformerContextCommon } from 'shiki'
+import type { Plugin } from 'unified'
+import { visit } from 'unist-util-visit'
+
+const TWOSLASH_POPUP_IMPORT_AST = {
+  type: 'mdxjsEsm',
+  data: {
+    estree: {
+      body: [
+        {
+          type: 'ImportDeclaration',
+          source: { type: 'Literal', value: 'nextra/components' },
+          specifiers: ['Popup', 'PopupTrigger', 'PopupPanel'].map(name => ({
+            type: 'ImportSpecifier',
+            imported: { type: 'Identifier', name },
+            local: { type: 'Identifier', name }
+          }))
+        }
+      ]
+    }
+  }
+} as any
+
+export const rehypeTwoslashPopup: Plugin<[], Root> = () => ast => {
+  // The tagName is being converted to lowercase when calling the shiki.codeToHtml method inside rehypePrettyCode.
+  // convert it back to Uppercase.
+  visit(ast, { tagName: 'popup' }, node => {
```

Comment:
I discovered a bug and have reported it. 
https://github.com/rehype-pretty/rehype-pretty-code/issues/251

Reply:
What if you put your plugin before rehypePretty code? Does it fix your issue?

Reply:
If I understand correctly, when running the Shiki transformers inside rehypePrettyCode, it creates the Popup Component's Node. 
So, if this rehype plugin is placed earlier, I think it won't work properly because there won't be any popup nodes in the hast yet.